spec/spec_helper.rb

```diff
@@ -29,3 +32,34 @@
 RSpec.configure do |c|
   c.default_facts = default_facts
 end
+
+# Stub the default dependency loader to do normal resolution except where overridden
+def stub_loader!
+  allow_any_instance_of(Puppet::Pops::Loader::DependencyLoader).to receive(:load).and_call_original
+end
+
+# Stub the dependency loader to resolve a named function to a double of some type
+#
+# @param name [String] The name of the puppet function to stub
+# @param dbl [RSpec double] Optional RSpeck double with `call` mocked;
+#        will be used instead of the block if supplied
+# @yield [*args] A block that takes all parameters given to the puppet function;
+#        required if dbl is not supplied, and ignored if dbl is supplied
+def stub_function(name, dbl = nil, &func)
+  func = dbl || func
+  stub = ->(_scope, *args, &block) do
+    func.call(*args, &block)
+  end
+  allow_any_instance_of(Puppet::Pops::Loader::DependencyLoader).to receive(:load).with(:function, name).and_return(stub)
+end
+
+def stub(name)
+  dbl = double(name)
+
+  yield dbl
+  stub_function(name, dbl)
+end
+
+def allow_call(dbl)
+  allow(dbl).to receive(:call)
+end
```

Comment:
With the invocations wrapped up into the shared context, I'm not sure that it's worth indirecting this. If we added it as an object method, I'd like it more. For example: 
```ruby
def stub_with_call(name)
  double(name).tap do |dbl|
    dbl.define_singleton_method(:allow_call) do
      allow(dbl).to receive(:call)
    end
    stub_function(name, dbl)
  end
end
```

I think that form is a little tidier because of the return rather than the block. Then downstream can do whatever with it as an imbued double. But it's pretty even either way.

Reply:
With this example, `allow` isn't within the right scope, and it fails. I had tried a variation on this (defining the method on the class) but it isn't obvious how to get the correct thing in there to call `allow` on.